### PR TITLE
Allow da.eye to support arbitrary chunking sizes with chunks='auto' 

### LINKED ISSUE #4635: Use da.eye with chunks="auto" fails
Currently if a user tries to create an array with `da.eye(..., chunks="auto")`, it fails. An MRE is included below.

<details>
<summary>Example:</summary>

```python
In [1]: import dask.array as da                                                 
da.ey	
In [2]: da.eye(100, chunks=10)                                                  
Out[2]: dask.array<eye, shape=(100, 100), dtype=float64, chunksize=(10, 10)>

In [3]: da.eye(100, chunks="auto")                                              
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-80d20afa2a92> in <module>
----> 1 da.eye(100, chunks="auto")

~/Developer/dask/dask/array/creation.py in eye(N, chunks, M, k, dtype)
    437     """
    438     if not isinstance(chunks, Integral):
--> 439         raise ValueError('chunks must be an int')
    440 
    441     token = tokenize(N, chunk, M, k, dtype)

ValueError: chunks must be an int
```

</details>

<br>

<details>
<summary>Environment:</summary>

```yaml
name: daskdev
channels:
  - conda-forge
  - defaults
dependencies:
  - appnope=0.1.0=py37_1000
  - asn1crypto=0.24.0=py37_1003
  - attrs=19.1.0=py_0
  - backcall=0.1.0=py_0
  - blas=1.1=openblas
  - bleach=3.1.0=py_0
  - bokeh=1.0.4=py37_1000
  - bzip2=1.0.6=h1de35cc_1002
  - ca-certificates=2019.3.9=hecc5488_0
  - certifi=2019.3.9=py37_0
  - cffi=1.12.2=py37h2d6ddff_1
  - chardet=3.0.4=py37_1003
  - click=7.0=py_0
  - cloudpickle=0.8.0=py_0
  - cryptography=2.6.1=py37hc2b1221_0
  - cycler=0.10.0=py_1
  - cytoolz=0.9.0.1=py37h470a237_1
  - dask-glm=0.1.0=0
  - dask-ml=0.12.0=py_0
  - decorator=4.3.2=py_0
  - defusedxml=0.5.0=py_1
  - distributed=1.26.0=py37_1
  - entrypoints=0.3=py37_1000
  - freetype=2.9.1=h597ad8a_1005
  - heapdict=1.0.0=py37_1000
  - idna=2.8=py37_1000
  - imageio=2.5.0=py37_0
  - ipykernel=5.1.0=py37h24bf2e0_1002
  - ipython=7.3.0=py37h24bf2e0_0
  - ipython_genutils=0.2.0=py_1
  - jedi=0.13.3=py37_0
  - jinja2=2.10=py_1
  - jpeg=9c=h1de35cc_1001
  - jsonschema=3.0.1=py37_0
  - jupyter_client=5.2.4=py_3
  - jupyter_core=4.4.0=py_0
  - kiwisolver=1.0.1=py37h04f5b5a_1002
  - libcxx=7.0.0=h2d50403_1
  - libffi=3.2.1=h0a44026_1005
  - libgfortran=3.0.1=0
  - libpng=1.6.36=ha441bb4_1000
  - libsodium=1.0.16=h1de35cc_1001
  - libtiff=4.0.10=h79f4b77_1001
  - llvm-meta=7.0.0=0
  - llvmlite=0.26.0=py37h8c7ce04_0
  - locket=0.2.0=py_2
  - markupsafe=1.1.1=py37h1de35cc_0
  - matplotlib-base=3.0.3=py37hf043ca5_0
  - mistune=0.8.4=py37h1de35cc_1000
  - msgpack-python=0.6.1=py37h04f5b5a_0
  - multipledispatch=0.6.0=py_0
  - nbconvert=5.4.1=py_2
  - nbformat=4.4.0=py_1
  - ncurses=6.1=h0a44026_1002
  - networkx=2.2=py_1
  - notebook=5.7.6=py37_0
  - numba=0.41.0=py37h1702cab_1000
  - numpy=1.16.2=py37_blas_openblash486cb9f_0
  - olefile=0.46=py_0
  - openblas=0.3.3=hdc02c5d_1001
  - openssl=1.1.1b=h1de35cc_1
  - packaging=19.0=py_0
  - pandas=0.24.2=py37h0a44026_0
  - pandoc=2.6=1
  - pandocfilters=1.4.2=py_1
  - parso=0.3.4=py_0
  - partd=0.3.9=py_0
  - pexpect=4.6.0=py37_1000
  - pickleshare=0.7.5=py37_1000
  - pillow=5.4.1=py37hbddbef0_1000
  - pip=19.0.3=py37_0
  - prometheus_client=0.6.0=py_0
  - prompt_toolkit=2.0.9=py_0
  - psutil=5.6.1=py37h1de35cc_0
  - ptyprocess=0.6.0=py37_1000
  - pycparser=2.19=py37_1
  - pygments=2.3.1=py_0
  - pyopenssl=19.0.0=py37_0
  - pyparsing=2.3.1=py_0
  - pyrsistent=0.14.11=py37h1de35cc_0
  - pysocks=1.6.8=py37_1002
  - python=3.7.1=hbdd33cc_1003
  - python-dateutil=2.8.0=py_0
  - pytz=2018.9=py_0
  - pywavelets=1.0.2=py37h917ab60_0
  - pyyaml=5.1=py37h1de35cc_0
  - pyzmq=18.0.1=py37h4cc6ddd_0
  - readline=7.0=hcfe32e1_1001
  - requests=2.21.0=py37_1000
  - scikit-image=0.14.2=py37h0a44026_1
  - scikit-learn=0.20.3=py37_blas_openblashc6dc708_0
  - scipy=1.2.1=py37_blas_openblash486cb9f_0
  - send2trash=1.5.0=py_0
  - setuptools=40.8.0=py37_0
  - six=1.12.0=py37_1000
  - sortedcontainers=2.1.0=py_0
  - sparse=0.6.0=py_0
  - sqlite=3.26.0=h1765d9f_1001
  - tblib=1.3.2=py_1
  - terminado=0.8.1=py37_1001
  - testpath=0.4.2=py37_1000
  - tk=8.6.9=ha441bb4_1000
  - toolz=0.9.0=py_1
  - tornado=6.0.1=py37h1de35cc_0
  - traitlets=4.3.2=py37_1000
  - urllib3=1.24.1=py37_1000
  - wcwidth=0.1.7=py_1
  - webencodings=0.5.1=py_1
  - wheel=0.33.1=py37_0
  - xz=5.2.4=h1de35cc_1001
  - yaml=0.1.7=h1de35cc_1001
  - zeromq=4.2.5=h0a44026_1006
  - zict=0.1.4=py_0
  - zlib=1.2.11=h1de35cc_1004
```

</details>

### PR DESCRIPTION
**Pull request by andersy005**
- [x] Tests added / passed
- [x] Passes `flake8 dask`

This Partially addresses #4635 in that it doesn't support asymmetric chunking

### CODE REVIEW

File: dask/array/creation.py [hunk 1]
```diff
@@ -411,7 +411,7 @@ def indices(dimensions, dtype=int, chunks='auto'):
     return grid
 
 
-def eye(N, chunks, M=None, k=0, dtype=float):
+def eye(N, chunks='auto', M=None, k=0, dtype=float):
```
Comment (jrbourbeau) [inline dask/array/creation.py]: Could you update the docstring for the `chunks` parameter to allow for values of type `str`

File: dask/array/tests/test_creation.py [hunk 2]
```diff
@@ -1,13 +1,10 @@
-import pytest
-pytest.importorskip('numpy')
-
+from dask.array.utils import assert_eq, same_keys, AxisError
```
[the remark this answers is not in the document]
Comment (andersy005) [reply dask/array/tests/test_creation.py]: I am just realizing that this is due to the fact that I used `autopep8 --aggressive` to fix flake8 issues. It ended up moving the imports around.  I will revert back to the original structure.

[the remark this answers is not in the document]
Comment (jrbourbeau) [reply dask/array/tests/test_creation.py]: Ah, I see! Thanks for the revert

File: dask/array/creation.py (hunk 1 above)
Comment (andersy005) [reply dask/array/creation.py]: Not sure the docstring update I added is clear enough :). Let me know if there's a better alternative.

**Comment by andersy005** [comment]
@jrbourbeau, thank you for reviewing this! Let me know if there are any additional changes needed.

**Comment by jrbourbeau** [comment]
Thanks for addressing my comments so quickly! I'm about to go offline for a little bit, but will do a final review in a few hours. 

> Do you know why travis isn't re-running checks on this PR after adding more commits to it?

My guess is it's related to https://www.githubstatus.com/incidents/nf4gcjtcsqdb

**Comment by TomAugspurger** [comment]
Thanks @andersy005! Does this fully close https://github.com/dask/dask/issues/4635? It looks like @lightsighter may have wanted something like

```
da.eye(12, chunks=(4, 3))
```

to work.

If so, we can still merge this PR because it's valuable as is, but we'll want to remove the "Fixes" in the original post so that the issue remains open.

File: dask/array/creation.py [hunk 3]
```diff
@@ -411,16 +411,17 @@ def indices(dimensions, dtype=int, chunks='auto'):
...
     """
     Return a 2-D Array with ones on the diagonal and zeros elsewhere.
 
     Parameters
     ----------
     N : int
       Number of rows in the output.
-    chunks: int
-        chunk size of resulting blocks
+    chunks: int, str
```
Comment (TomAugspurger) [inline dask/array/creation.py]: Copying from `from_array`, we may want

```
    chunks : int, str
        How to chunk the array. Must be one of the following forms:
        -   A blocksize like 1000.
        -   A size in bytes, like "100 MiB" which will choose a uniform
            block-like shape
        -   The word "auto" which acts like the above, but uses a configuration
            value ``array.chunk-size`` for the chunk size
```

I removed the tuple options, as I don't think those will work yet.

Comment (andersy005) [reply dask/array/creation.py]: Thank you for the suggestion, @TomAugspurger! I will update the docstring to this new version.

**Comment by andersy005** [comment]
@TomAugspurger, 

Regarding 

```
da.eye(12, chunks=(4, 3))
```
I will take a stab at it, and see if I can get it to work as part of this PR.

File: dask/array/creation.py (hunk 3 above)
Comment (jrbourbeau) [reply dask/array/creation.py]: Good suggestion @TomAugspurger. Thanks for handling this @andersy005

**Comment by jakirkham** [comment]
I think @lightsighter is more interested in `da.eye(12, chunks="auto")` than anisotropic chunking. Though I suppose if `"auto"` resulted in anisotropic chunking, then he would want that indirectly.

**Comment by TomAugspurger** [comment]
To be clear, I have no objections to merging this as is (well, once the docstring is updated).

@lighsighter can you confirm that something like `chunks='auto'` suffices for you? Or were you also hoping for something like `da.eye(12, chunks=(4, 3))`.

**Comment by jakirkham** [comment]
@lightsighter, if you are around, would you have a chance to answer Tom's question ([above]( https://github.com/dask/dask/pull/4834#issuecomment-495293733 )) and see if this PR solves the issue you encountered?

**Comment by lightsighter** [comment]
I really wanted both so I could compare performance. I would have a preference though for 'chunks=(4,3)' over 'chunks="auto"' though because I really ultimately want the ability to pick my chunk sizes like I can for other Dask operations.

**Comment by TomAugspurger** [comment]
Good to know. @andersy005 do you want to look into that? If so, do you prefer to do it here or in a separate PR?

**Comment by andersy005** [comment]
@TomAugspurger, it appears that getting `da.eye(12, chunks=(4, 3))` to work is going to take me longer than I expected. As a quick/naive prototype, I attempted making some changes to  

https://github.com/dask/dask/blob/2f308e72a46de818055819108429e8f7ce7ba03b/dask/array/creation.py#L455-L462

```python

    for i, vchunk in enumerate(vchunks):
        for j, hchunk in enumerate(hchunks):
            if (j - i - 1) * vchunk <= k <= (j - i + 1) * hchunk:
                eye[name_eye, i, j] = (np.eye, vchunk, hchunk, k - (j - i) * vchunk, dtype)
            else:
                eye[name_eye, i, j] = (np.zeros, (vchunk, hchunk), dtype)
```
However, as you can see in examples below, my logic is broken :).  I need some guidance on what changes will need to be made in the original `for loop` (above) when `chunks` is not the same across both dimensions like in `da.eye(12, (4, 3))`:

```python
In [1]: import dask.array as da                                                                                                                           

In [2]: da.eye(10, 'auto')                                                                                                                                
Out[2]: dask.array<eye, shape=(10, 10), dtype=float64, chunksize=(10, 10)>

In [3]: da.eye(12, (4, 3))                                                                                                                                
Out[3]: dask.array<eye, shape=(12, 12), dtype=float64, chunksize=(4, 3)>

In [4]: da.eye(10, 'auto').compute()                                                                                                                      
Out[4]: 
array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])

In [5]: da.eye(12, (4, 3)).compute()                                                                                                                      
Out[5]: 
array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [6]: da.eye(10, (3, 3)).compute()                                                                                                                      
Out[6]: 
array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 1.]])
```

**Comment by TomAugspurger** [comment]
Hmm, it's not immediately clear to me how to handle asymmetric chunking... I would need to think about it more.

**Comment by andersy005** [comment]
No worries. Would it be okay to leave this PR open for a few days as we figure out how to handle the asymmetric chunking? In the meantime, keep me posted if you figure it out ( I too will keep looking into it more )

**Comment by TomAugspurger** [comment]
Yep, that's not a problem.

**Comment by jakirkham** [comment]
I would propose we merge this as is and deal with anisotropic chunking in a follow up PR. As it is this already fixes part of the issue and is ready to go. It won’t hinder later work from occurring and might make it easier as we can deal with the changes and their consequences independently. Thoughts?

**Comment by andersy005** [comment]
@jakirkham, your suggestion sounds good to me

**Comment by andersy005** [comment]
@TomAugspurger, I updated the PR's title and removed the "Fixes" in the original comment so that the issue remains open. Feel free to merge this.

**Comment by TomAugspurger** [comment]
Thanks @andersy005!

**Comment by jakirkham** [comment]
Great, thanks all! Glad to see this nice improvement in. 😄

**Pull request merged**